src/inmanta/server/services/userservice.py

```diff
@@ -0,0 +1,125 @@
+
+    def __init__(self) -> None:
+        super().__init__(SLICE_USER)
+
+    def get_dependencies(self) -> list[str]:
+        return [SLICE_DATABASE]
+
+    def get_depended_by(self) -> list[str]:
+        return [SLICE_TRANSPORT]
+
+    @protocol.handle(protocol.methods_v2.list_users)
+    async def list_users(self) -> list[model.User]:
+        return [user.to_dao() for user in await data.User.get_list()]
+
+    @protocol.handle(protocol.methods_v2.add_user)
+    async def add_user(self, username: str, password: str) -> model.User:
+        verify_authentication_enabled()
+        if not password or len(password) < 8:
+            raise exceptions.BadRequest("the password should be at least 8 characters long")
+
+        # hash the password
+        pw_hash = nacl.pwhash.str(password.encode())
+
+        # insert the user
+        try:
+            user = data.User(
+                username=username,
+                password_hash=pw_hash.decode(),
+                auth_method=AuthMethod.DATABASE.value,
+            )
+            await user.insert()
+        except asyncpg.UniqueViolationError:
+            raise exceptions.Conflict(f"A user with name {username} already exists.")
+        return user.to_dao()
+
+    @protocol.handle(protocol.methods_v2.delete_user)
+    async def delete_user(self, username: str) -> None:
+        verify_authentication_enabled()
+        user = await data.User.get_one(username=username)
+        if user is None:
+            raise exceptions.NotFound(f"User with name {username} does not exist.")
+
+        await user.delete()
+
+    @protocol.handle(protocol.methods_v2.set_password)
+    async def set_password(self, username: str, password: str) -> None:
+        verify_authentication_enabled()
+        if not password or len(password) < 8:
+            raise exceptions.BadRequest("the password should be at least 8 characters long")
+        # check if the user already exists
+        user = await data.User.get_one(username=username)
+        if not user:
+            raise exceptions.NotFound(f"User with name {username} does not exist.")
+
+        # hash the password
+        pw_hash = nacl.pwhash.str(password.encode())
+
+        # insert the user
+        await user.update_fields(password_hash=pw_hash.decode())
+
+    @protocol.handle(protocol.methods_v2.login)
+    async def login(self, username: str, password: str) -> common.ReturnValue[model.LoginReturn]:
+        verify_authentication_enabled()
+        # check if the user exists
+        user = await data.User.get_one(username=username)
+        if not user or not user.password_hash:
+            raise exceptions.UnauthorizedException()
+        try:
+            nacl.pwhash.verify(user.password_hash.encode(), password.encode())
+        except nacl.exceptions.InvalidkeyError:
+            raise exceptions.UnauthorizedException()
+
+        token = common.encode_token([str(const.ClientType.api.value)], expire=None)
```

Comment:
Is it good practice to never expire this? I honestly don't know mysefl.

Reply:
there is a follow up ticket for that: https://app.zenhub.com/workspaces/core-5a7c152b04a1652024289b7b/issues/gh/inmanta/inmanta-core/5643

Reply:
Ok, that'll do!